### ISSUE TITLE
`cargo hakari generate` crashed

### ISSUE THREAD

**Issue by smndtrl** [guppy-rs/guppy#255]
When trying out `hakari`, I couldn't run the `generate` command.
Let me know if I can provide more info.

```
$ cargo hakari generate
The application panicked (crashed).
Message:  index out of bounds: the len is 7282 but the index is 7282
Location: ~/.cargo/registry/src/index.crates.io-6f17d22bba15001f/guppy-0.17.5/src/petgraph_support/topo.rs:49

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ BACKTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
   1: color_eyre::config::PanicHook::into_panic_hook::{{closure}}::hbabd9e3254037d94
      at <unknown source file>:<unknown line>
   2: std::panicking::rust_panic_with_hook::hfd477be88114f3c4
      at <unknown source file>:<unknown line>
   3: std::panicking::begin_panic_handler::{{closure}}::hb957e98065a4568c
      at <unknown source file>:<unknown line>
   4: std::sys_common::backtrace::__rust_end_short_backtrace::ha9a82577e2e3f0de
      at <unknown source file>:<unknown line>
   5: rust_begin_unwind<unknown>
      at <unknown source file>:<unknown line>
   6: core::panicking::panic_fmt::h281c4bbb02f68665
      at <unknown source file>:<unknown line>
   7: core::panicking::panic_bounds_check::h975aed5e94c0faa6
      at <unknown source file>:<unknown line>
   8: guppy::petgraph_support::topo::TopoWithCycles<Ix>::new::h96c0455e22f74fde
      at <unknown source file>:<unknown line>
   9: once_cell::imp::OnceCell<T>::initialize::{{closure}}::ha03be1ce336e21ab
      at <unknown source file>:<unknown line>
  10: once_cell::imp::initialize_or_wait::hd5d8d656a2a768a3
      at <unknown source file>:<unknown line>
  11: once_cell::imp::OnceCell<T>::initialize::h49bb92dc6e588fbb
      at <unknown source file>:<unknown line>
  12: guppy::graph::feature::resolve::FeatureSet::root_ids::he4960f4645d3e97f
      at <unknown source file>:<unknown line>
  13: <alloc::collections::btree::map::ExtractIf<K,V,F,A> as core::iter::traits::iterator::Iterator>::next::hd7438f48ef318cf5
      at <unknown source file>:<unknown line>
  14: hakari::hakari::Hakari::build::he98dd6651e40e84a
      at <unknown source file>:<unknown line>
  15: cargo_hakari::command::Command::exec::hdc1c66fe7fdfa045
      at <unknown source file>:<unknown line>
  16: cargo_hakari::main::hf4a74d35af379ec9
      at <unknown source file>:<unknown line>
  17: std::sys_common::backtrace::__rust_begin_short_backtrace::h5d74b18ad36c9ab3
      at <unknown source file>:<unknown line>
  18: std::rt::lang_start::{{closure}}::h0234ee382a08a736
      at <unknown source file>:<unknown line>
  19: std::rt::lang_start_internal::h7d9fe2c277f41ab2
      at <unknown source file>:<unknown line>
  20: main<unknown>
      at <unknown source file>:<unknown line>
  21: __libc_start_main<unknown>
      at <unknown source file>:<unknown line>
  22: _start<unknown>
      at <unknown source file>:<unknown line>
```

**Comment by sunshowers** [comment]
Thanks for the report. That's really strange -- it looks like it's coming from this line:

https://github.com/guppy-rs/guppy/blob/dd968ee0b08f046ec753e8f318f9b0bd2d043f0f/guppy/src/petgraph_support/topo.rs#L48-L50

As the comment above says, that shouldn't happen because the graph is _compact_ (ensured via the marker trait): all indexes must be between 0 and the number of nodes in the graph. A topo sort of the graph retains all the nodes in the graph, so it's very strange that this is failing at all.

The graph being used is here:

https://github.com/guppy-rs/guppy/blob/dd968ee0b08f046ec753e8f318f9b0bd2d043f0f/guppy/src/graph/feature/graph_impl.rs#L231-L244

This graph is compact, because:

1. The parent graph (self.dep_graph()) is compact, and
2. While edges are filtered out, nodes are not.

Additionally, I've used hakari on a number of repos and have never seen anything like this before.

I don't really have time at the moment to try and debug this in more detail, and I also don't have access to your workspace. Would you be open to checking out guppy and adding some debug logging to these lines to see what's going on? This is very strange.

**Comment by sunshowers** [comment]
I think one thing you can do is to, instead of writing to a Vec, build a BTreeMap -- then see if there are any gaps in the numbers. You can then try mapping those numbers back to particular nodes to see if there's anything special going on.

**Comment by sunshowers** [comment]
Addressed by #288. Thanks @stefanboere!

**Comment by sunshowers** [comment]
Actually I'm not really convinced by the fix for #288 -- having looked at it more, it just isn't possible for a graph that its topo sort results in fewer nodes than the graph itself. I think I'm going to have to experiment with this further to figure out what's going on.

**Comment by sunshowers** [comment]
The fix for this is now out in cargo-hakari 0.9.30.
